lib/accesslists/validate.go

```diff
@@ -28,74 +28,93 @@ import (
 
 // ValidateAccessListWithMembers validates a new or existing AccessList with a list of AccessListMembers.
 func ValidateAccessListWithMembers(ctx context.Context, accessList *accesslist.AccessList, members []*accesslist.AccessListMember, g AccessListAndMembersGetter) error {
+	if err := validateAccessListNesting(ctx, accessList, members, g); err != nil {
+		return trace.Wrap(err)
+	}
+	return nil
+}
+
+// validateAccessListNesting validates if nested AccessList owners and members meet max depth (of
+// 10) requirement and don't create cycles.
+func validateAccessListNesting(ctx context.Context, accessList *accesslist.AccessList, members []*accesslist.AccessListMember, g AccessListAndMembersGetter) error {
 	for _, owner := range accessList.Spec.Owners {
-		if owner.MembershipKind != accesslist.MembershipKindList {
-			continue
-		}
-		ownerList, err := g.GetAccessList(ctx, owner.Name)
-		if err != nil {
-			return trace.Wrap(err)
-		}
-		if err := validateAddition(ctx, accessList, ownerList, RelationshipKindOwner, g); err != nil {
+		if err := validateAccessListOwner(ctx, accessList, owner, g); err != nil {
 			return trace.Wrap(err)
 		}
 	}
 	for _, member := range members {
-		if member.Spec.MembershipKind != accesslist.MembershipKindList {
-			continue
-		}
-		memberList, err := g.GetAccessList(ctx, member.GetName())
-		if err != nil {
-			return trace.Wrap(err)
-		}
-		if err := validateAddition(ctx, accessList, memberList, RelationshipKindMember, g); err != nil {
+		if err := ValidateAccessListMember(ctx, accessList, member, g); err != nil {
 			return trace.Wrap(err)
 		}
 	}
 	return nil
 }
 
-// ValidateAccessListMember validates a new or existing AccessListMember for an Access List.
+// ValidateAccessListMember validates AccessListMember. That includes nested AccessList members
+// validation.
 func ValidateAccessListMember(
 	ctx context.Context,
 	parentList *accesslist.AccessList,
 	member *accesslist.AccessListMember,
 	g AccessListAndMembersGetter,
 ) error {
-	if member.Spec.MembershipKind != accesslist.MembershipKindList {
-		return nil
+	if err := validateAccessListMemberBasic(member); err != nil {
+		return trace.Wrap(err)
+	}
+	if err := validateAccessListMemberOrOwnerNesting(ctx, parentList, member.GetName(), RelationshipKindMember, member.Spec.MembershipKind, g); err != nil {
+		return trace.Wrap(err)
+	}
+	return nil
+}
+
+// validateAccessListMemberBasic performs basic fields validation for AccessListMember.
+func validateAccessListMemberBasic(member *accesslist.AccessListMember) error {
```

Comment:
When are we supposed to use `validateAccessListMemberBasic` vs. `ValidateAccessListMember`? Can we have that explained in the comment

Reply:
Not sure if you spotted `validateAccessListMemberBasic` is private. This indicates it can't be used outside the package. Do you think it needs further explanations?